core/src/main/java/org/apache/calcite/rel/rules/FilterProjectTransposeRule.java

```diff
@@ -181,6 +183,14 @@ public void onMatch(RelOptRuleCall call) {
                 .project(project.getProjects(), project.getRowType().getFieldNames())
                 .build();
 
+    // if new filter is always false and none of it's inputs use
+    // correlation variables - simplify new project to empty values
+    if (newCondition.isAlwaysFalse() && newFilterRel.getInputs().stream()
+        .map(RelOptUtil::getVariablesUsed)
+        .allMatch(Collection::isEmpty)) {
+      newProjRel = relBuilder.push(newProjRel).empty().build();
```

Comment:
If I am not mistaken, this simplification would have already taken place in the "else" block (copyFilter=false, line 175), because the RelBuilder#filter method already performs it, so it seems kind of redundant in that case. I think we would only need to add it to the "if" block (copyFilter=true).
Now, the problem is that I'm not sure we could use this simplification (which uses the relBuilder) in the case of copyFilter, as stated in the javadoc description of the private simplifyFilterCondition method.
There was some discussion about this rule and the condition simplification in https://issues.apache.org/jira/browse/CALCITE-2582 ; maybe @zabetak , who in the end committed that fix, can help.

Reply:
I thought behind exactly the same lines

Reply:
The rules that are using the copy bring some extra headache and many of them end up creating problems but let's not open this discussion again here. 

Now regarding the change, I think it is not safe to do it here. If our input was an `EnumerableFilter` and `EnumerableProject` then the way is the code right now we may end up creating an `EnumerableProject` over a `LogicalValues` which may end up being worse than an `EnumerableFilter` that is always false. 

Given that the code even before this change retained a filter that is always false in the case of copying I would suggest to keep this behavior in order to advance faster this PR. It may be better to discuss the removal of the filter under another issue.

Reply:
yes @rubenada , I see your point with the ```copyFilter``` if. I would not want to add the chunk:
```java
call.transformTo(...)
return; 
```
inside this ```if (copyFilter)```, and alternative introduction of another boolean variable to check it before ```copyProject``` also seems bad to me. I'll look for better way of doing this, thanks.

Reply:
Hi @zabetak , thanks for your quick reply. Leaving the Filter with false condition causes test failures for some adapters, like MongoDB, because such filter can't be implemented properly.

Reply:
> like MongoDB, because such filter can't be implemented properly.

That is exactly the "cost of support" for an adapter. That is why I am not very keen on merging new adapters (e.g. https://github.com/apache/calcite/pull/1591 )

There are the following approaches:
a) Implement a consistent way to ensure `where false` is never produced. I guess that rule does not belong to FilterProjectTransposeRule. `where false` could appear in many places. And there might be a generic rule that replaces `where false` with an empty relation, so "adapters that do not support `where false` could use the rule". I guess such a rule should already exist because it sounds like a common optimization
b) Implement `where filter` logic for `mongodb`. I do recognize your fix has nothing to do with `mongo`, but that is exactly the price we pay for having lots of adapters
c) De-support `mongodb` adapter. In other words, drop the code from Git

Reply:
> like MongoDB, because such filter can't be implemented properly.

That is exactly the "cost of support" for an adapter. That is why I am not very keen on merging new adapters (e.g. https://github.com/apache/calcite/pull/1591 )

There are the following approaches:
a) Implement a consistent way to ensure `where false` is never produced. I guess that rule does not belong to FilterProjectTransposeRule. `where false` could appear in many places. And there might be a generic rule that replaces `where false` with an empty relation, so "adapters that do not support `where false` could use the rule". I guess such a rule should already exist because it sounds like a common optimization
b) Implement `where filter` logic for `mongodb`. I do recognize your fix has nothing to do with `mongo`, but that is exactly the price we pay for having lots of adapters
c) De-support `mongodb` adapter. In other words, drop the code from Git

Reply:
I think option a) is the way to go and the rule to use is most likely `FilterReduceExpressionsRule`. I think like this we can also get rid of the `simplifyFilterCondition` method.

Reply:
Thank you guys for good advice, I also like most the option **a** .

Reply:
Can you take
```
newFilterRel.getInputs().stream()
        .map(RelOptUtil::getVariablesUsed)
        .allMatch(Collection::isEmpty)
```
and replace it with a call to a new method `RelOptUtil.getVariablesUsed(filter.getInputs())`.

Also, "its" not "it's".

Reply:
I don't agree with this talk of dropping adapters. (Though of course we should not accept 'bad' code that will have a high cost of upkeep.) It isn't about whether the adapter supports 'where false'. It is about whether the adapter implements relational algebra.

When applying transformation rules, we are reasoning in relational algebra, not in the logic of the target adapter. We only have to reason in the logic of the target adapter when generating the code (query language) of that adapter. That reduces the amount of code that lives in the adapter; most code is shared among all adapters.

Reply:
Hello @zabetak , 

Activation of ```ReduceExpressionsRule.FILTER_INSTANCE``` causes a bunch of mysterious [test failures](https://drive.google.com/open?id=10_th-enGEFc5c6ei0B-TflStC8FaW7ze). For example, ```EnumerableCorrelateTest.semiJoinCorrelateWithFilterCorrelateRule``` failed only because the rule was active, though it was never actually applied. It also looks very strange that this rule along with other reduction rules wasn't enabled by default in [RelOptUtil.registerDefaultRules(...)](https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/plan/RelOptUtil.java#L1941).

Reply:
I had a quick look in `EnumerableCorrelateTest` and it seems that it fails because the plan is slightly different but from what I see still semantically correct. Plan comparison is not that clever so it is common that you need to update the expected plan when such cases occur. 

Other than that it does not really matter if the rule produced a new expression or not. Adding a new rule may also change the order that other rules are matched leading to changes in the expected plan which might explain why tests fail.

Reply:
Even though the code moved to another place can you explain again why we cannot perform the simplification if the input contains correlation variables? It seems that @vlsi also raised this question but I didn't understand your answer. If the filter is always false then why do we need to look into the variables?

Reply:
Ok, a closer look at [this comment](https://github.com/apache/calcite/pull/1610#discussion_r353628674) shows that early simplification of filter to ```EnumerableValues(tuples=[[]])``` prevents calcite from throwing out correlation join left side and table scan is present in plan when it's absolutely unnecessary.

Reply:
> prevents calcite from throwing out correlation join left side and table scan is present in plan when it's absolutely unnecessary.

Should it be covered by a different rule?
In other words, "if correlation other side is always empty, then the correlation result is empty".

Reply:
@vlsi , @zabetak :) I already don't know, magically the case was fixed by rebasing on master.